商品出品機能

ganaha-mikoto opened the pull request:
# What
商品出品機能のためのMVCとルーティング
販売手数料のJavaScript
モデル単体テスト


# Why
商品を適切な情報を入力すると保存できるようするため


ログイン状態の場合は、商品出品ページへ遷移できる動画
[https://gyazo.com/8b0483240e27223ac1e69dee4d68c04f](url)
価格が入力されると同時に、販売手数料と販売利益が表示される動画
[https://gyazo.com/71c0d443658556ec98afb05b614fa1ba](url)
必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
[https://gyazo.com/613458a1e99ff40cb6946e6d64510c78](url)
入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
[https://gyazo.com/ca330a513e6761f22b310168f0590593](url)
ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
[https://gyazo.com/81f3147a2ebf18b5f98b33c1665f7dea](url)
テスト結果の画像
[https://gyazo.com/ade6386bcab55b6c1ebb47fcf9e202bf](url)

ganaha-mikoto committed:
データの登録

ルーティング
コントローラーの設定

ganaha-mikoto committed:
画像投稿機能

Gem fileの編集
config/application.rbのminimagicの導入の設定
active_storageのinstallとmigrate

ganaha-mikoto committed:
ActiveHash

モデルにカテゴリー、状態、配送料、配送場所、配送日時をアクティブハッシュを継承し設定

ganaha-mikoto committed:
アクティブハッシュのバリデーション

ganaha-mikoto committed:
ログインしていない場合のリダイレクト

コントローラーにログインしていない時にログイン画面へ移行するようにした

ganaha-mikoto committed:
collectionのhogeを修正、エラーメッセージのインスタンスの引き継ぎ

ganaha-mikoto committed:
JavaScriptの設定

ganaha-mikoto committed:
微調整

エラーが出たため言語・時間の設定を削除
出品ボタンにリンク生成
不要なアソシエーションを削除
データベースの更新

ganaha-mikoto committed:
エラーメッセージを一つずつ表示するように設定

ganaha-mikoto committed:
テスト

yuri240 commented on app/javascript/product_price.js:
```diff
@@ -0,0 +1,23 @@
+  if (priceInput) {
+    priceInput.addEventListener("input", () => {
+      const inputValue = priceInput.value;
+      console.log(inputValue);
```
デバック用の`console.log(inputValue);`は削除しましょう！

理由：企業提出時には不要となるためです。

yuri240 commented on app/models/product.rb:
```diff
@@ -1,6 +1,115 @@
+      { id: 6, name: '傷や汚れあり' },
+      { id: 7, name: '全体的に状態が悪い' }
+    ]
+  end
```
アクティブハッシュに関する記述ですが、
`item.rb`とは別で5つモデルを作成し、
そちらの記載するようにしましょう！

例）`app/models/category.rb`
```ruby
  class Category < ActiveHash::Base
    self.data = [
      { id: 1, name: '---' },
      { id: 2, name: 'メンズ' },
      { id: 3, name: 'レディース' },
      { id: 4, name: 'ベビー・キッズ' },
      { id: 5, name: 'インテリア・住まい・小物' },
      { id: 6, name: '本・音楽・ゲーム' },
      { id: 7, name: 'おもちゃ・ホビー・グッズ' },
      { id: 8, name: '家電・スマホ・カメラ' },
      { id: 9, name: 'スポーツ・レジャー' },
      { id: 10, name: 'ハンドメイド' },
      { id: 11, name: 'その他' }
    ]
  end
```

yuri240 commented on app/models/product.rb:
```diff
@@ -1,6 +1,115 @@
+
+  validates :name, :description, :category_id, :condition_id, :shipping_charge_id, :shipping_area_id, :shipping_day_id, :price, presence: true
+  validates :category_id, :condition_id, :shipping_charge_id, :shipping_area_id, :shipping_day_id, numericality: { other_than: 1 , message: "can't be blank"}
+  validates :price, numericality: { greater_than: 0, less_than_or_equal_to: 9_999_999, message: "must be between 1 and 9999999" }
```
`price`カラムは、整数のみ保存が行われるようバリデーションを追加しましょう！

理由：
現状の記述ですと、小数の販売価格が保存できてしまうため。

yuri240 replied:
` validates :price, numericality: { greater_than: 0, less_than_or_equal_to: 9_999_999, message: "must be between 1 and 9999999" }`

300円以上でないと登録できないようバリデーションを修正しましょう！

yuri240 commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,81 @@
+      expect(@product.errors.full_messages).to include("Price must be between 1 and 9999999")
+    end
+
+    it '価格が0以下では登録できない' do
```
実装要件に合わせて、`価格が300以下では登録できない`に修正しましょう！

yuri240 commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,81 @@
+    end
+  end
+
+  context '商品の新規登録ができない場合' do
```
商品にはユーザー情報が紐づくことを確認するために、以下のテストを追加しましょう。
```
・ユーザー情報がない場合は登録できないこと
```

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
コメントさせていただきましたので、ご確認お願いいたします。


```ruby
◎ 次回提出時のお願い ◎
お手数ではありますがテストコードを追加した後は、
改めてテストコード実行後のキャプチャ画像を撮影いただき、
こちらのプルリクエストへ添付していただくようお願いいたします。
```

ganaha-mikoto committed:
モデルテストとモデルの修正

JavaScriptのコンソールの削除
アクティブハッシュを別モデルに分ける
ユーザーと紐付き関連のテストを作る
config/application.rbにアクティブハッシュモデルたちが読み込まれるよう設定をたす

新たなテスト成功の画像
[https://gyazo.com/0c812288595623a4ce9e9653bb307592](url)

koyamatakumi commented on app/models/product.rb:
```diff
@@ -1,6 +1,17 @@
+
+  validates :name, :description, :category_id, :condition_id, :shipping_charge_id, :shipping_area_id, :shipping_day_id, :price, presence: true
+  validates :category_id, :condition_id, :shipping_charge_id, :shipping_area_id, :shipping_day_id, numericality: { other_than: 1, message: "can't be blank" }
+  validates :price, numericality: { only_integer: true, greater_than: 300, less_than_or_equal_to: 9_999_999, message: "must be between 300 and 9999999" }
```
```ruby
  validates :price, numericality: { only_integer: true, greater_than: 300, less_than_or_equal_to: 9_999_999, message: "must be between 300 and 9999999" }
```
こちらのバリデーションですと、最低出品価格は301円となってしまっています。
300円の商品も出品できるようにしましょう。

koyamatakumi commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,87 @@
+      @product.price = 10000000
+      @product.valid?
+      expect(@product.errors.full_messages).to include("Price must be between 300 and 9999999")
+    end
```
「画像が紐づいていない場合登録できない」のテストも追加しましょう。

koyamatakumi reviewed (commented):
ganaha-mikoto様
レビュー依頼フォームの送信ありがとうございます。 修正いただいた内容を確認させていただき、追加でコメントいたしましたのでご確認お願いいたします。
修正後はコミット、プッシュを忘れずに行い、改めてレビュー依頼フォームの送信をお願い致します。着実にゴールに近づいています。ラストスパート頑張っていきましょう！

ganaha-mikoto committed:
画像保存に関するバリデーションとテスト

画像保存に関するバリデーションとテスト
値段を300円でも出品できるようにした

テスト画像
[https://gyazo.com/96db5268699c21d484968bf3ea1b816b](url)

div-reviewer commented on app/javascript/product_price.js:
```diff
@@ -0,0 +1,21 @@
+      const fee = Math.floor(inputValue * 0.1);
+      
+      // 販売利益を計算
+      const profit = inputValue - fee;
```
こちらにもMath.floorを追加しましょう。

div-reviewer reviewed (commented):
ganaha-mikoto様、学習お疲れさまです。
商品出品の再レビュー依頼ありがとうございました。

前回のコメントにあった修正点が綺麗に直せましたね。素晴らしいです。
こちら`LGTM`とさせて頂きます。実装お疲れさまでした。

**【修正について】**
修正して頂ける場合は`再度レビュー依頼は出さなくて大丈夫`です。
修正後、コミットとプッシュは忘れずにお願いします。

**【こちらは次回の機能に入る際の注意点です】**
プルリクエストのMerge、ローカルのmainブランチプルを忘れずに行い、ブランチを切ってから次の実装に取り組んで頂ければと思います。

**ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！**
今後の実装も応援しております！どうぞよろしくおねがいします。
